File: backend/src/utils/number_format.py
"""
Number Format Utility — Per-client currency formatting (mirrors timezone pattern).

Usage:
    from ..utils.number_format import get_client_currency_code, format_currency

    currency = get_client_currency_code(supabase, client_id)   # e.g. "AUD"
    label    = format_currency(123456.78, currency)             # "$123,457"
    label    = format_currency(123456.78, currency, decimals=2) # "$123,456.78"

Supported ISO 4217 codes and their display symbols:

    AUD  →  $     (Australian Dollar)
    USD  →  $     (US Dollar)
    NZD  →  $     (New Zealand Dollar)
    CAD  →  $     (Canadian Dollar)
    GBP  →  £     (British Pound)
    EUR  →  €     (Euro)
    SGD  →  $     (Singapore Dollar)
    HKD  →  $     (Hong Kong Dollar)
    JPY  →  ¥     (Japanese Yen — no decimals)
    INR  →  ₹     (Indian Rupee)

All dollar-sign currencies are disambiguated by their code label when needed in AI
prompts (e.g. "AUD $123,456") so the LLM understands the context.
"""

import logging
from typing import Optional

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Currency metadata
# ---------------------------------------------------------------------------
_CURRENCY_META = {
    "AUD": {"symbol": "$", "code_label": "AUD", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "USD": {"symbol": "$", "code_label": "USD", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "NZD": {"symbol": "$", "code_label": "NZD", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "CAD": {"symbol": "$", "code_label": "CAD", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "SGD": {"symbol": "$", "code_label": "SGD", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "HKD": {"symbol": "$", "code_label": "HKD", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "GBP": {"symbol": "£", "code_label": "GBP", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "EUR": {"symbol": "€", "code_label": "EUR", "decimals": 0, "thousands_sep": ".", "decimal_sep": ","},
    "JPY": {"symbol": "¥", "code_label": "JPY", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
    "INR": {"symbol": "₹", "code_label": "INR", "decimals": 0, "thousands_sep": ",", "decimal_sep": "."},
}

_DEFAULT_CURRENCY = "AUD"


# ---------------------------------------------------------------------------
# Client currency loader (same pattern as _get_client_timezone)
# ---------------------------------------------------------------------------
def get_client_currency_code(supabase, client_id: Optional[str]) -> str:
    """
    Load per-client currency code from the clients table.

    Falls back to AUD if the column is missing, null, or the DB call fails.
    Mirrors _get_client_timezone() in am_efficiency_analyzer.py.
    """
    if not client_id:
        return _DEFAULT_CURRENCY
    try:
        resp = (
            supabase.table("clients")
            .select("currency_code")
            .eq("id", client_id)
            .limit(1)
            .execute()
        )
        rows = resp.data or []
        code = (rows[0].get("currency_code") or _DEFAULT_CURRENCY) if rows else _DEFAULT_CURRENCY
        return code.upper().strip() if code else _DEFAULT_CURRENCY
    except Exception:
        return _DEFAULT_CURRENCY


# ---------------------------------------------------------------------------
# Formatter
# ---------------------------------------------------------------------------
def format_currency(
    amount,
    currency_code: str = _DEFAULT_CURRENCY,
    decimals: Optional[int] = None,
    include_code: bool = False,
) -> str:
    """
    Format a numeric amount as a currency string.

    Args:
        amount:        Numeric value (int / float / None). None → "N/A".
        currency_code: ISO 4217 code (e.g. "AUD"). Unknown codes fall back to AUD.
        decimals:      Override decimal places (default: from currency meta).
        include_code:  Prefix with code label, e.g. "AUD $123,456". Useful in AI
                       prompts so the LLM knows which currency is intended.

    Returns:
        Formatted string, e.g. "$123,456" or "AUD $1,234.56" or "N/A".
    """
    if amount is None:
        return "N/A"

    meta = _CURRENCY_META.get(currency_code.upper(), _CURRENCY_META[_DEFAULT_CURRENCY])
    dp = decimals if decimals is not None else meta["decimals"]
    symbol = meta["symbol"]
    code_label = meta["code_label"]

    try:
        value = float(amount)
    except (TypeError, ValueError):
        return "N/A"

    # Format with Python's built-in comma thousands + configurable decimals
    formatted = f"{value:,.{dp}f}"

    result = f"{symbol}{formatted}"
    if include_code:
        result = f"{code_label} {result}"
    return result


def format_currency_range(low, high, currency_code: str = _DEFAULT_CURRENCY) -> str:
    """Format a value range, e.g. '$50,000 – $100,000'."""
    return f"{format_currency(low, currency_code)} – {format_currency(high, currency_code)}"

File: frontend/src/utils/numberFormat.ts
/**
 * Number Format Utility — per-client currency formatting.
 *
 * Mirrors the backend number_format.py pattern: reads currency_code from the
 * client record and formats financial figures consistently throughout the UI.
 *
 * Usage:
 *   import { formatCurrency, formatCurrencyCompact } from '../utils/numberFormat';
 *
 *   formatCurrency(123456, 'AUD')           // "$123,456"
 *   formatCurrency(123456.78, 'AUD', 2)     // "$123,456.78"
 *   formatCurrencyCompact(1234567, 'AUD')   // "$1.2M"
 */

/** Supported ISO 4217 currency codes and their display metadata. */
const CURRENCY_META: Record<string, { symbol: string; locale: string }> = {
  AUD: { symbol: '$', locale: 'en-AU' },
  USD: { symbol: '$', locale: 'en-US' },
  NZD: { symbol: '$', locale: 'en-NZ' },
  CAD: { symbol: '$', locale: 'en-CA' },
  SGD: { symbol: '$', locale: 'en-SG' },
  HKD: { symbol: '$', locale: 'en-HK' },
  GBP: { symbol: '£', locale: 'en-GB' },
  EUR: { symbol: '€', locale: 'de-DE' },
  JPY: { symbol: '¥', locale: 'ja-JP' },
  INR: { symbol: '₹', locale: 'en-IN' },
};

const DEFAULT_CURRENCY = 'AUD';

function getMeta(currencyCode: string) {
  return CURRENCY_META[currencyCode?.toUpperCase()] ?? CURRENCY_META[DEFAULT_CURRENCY];
}

/**
 * Format a number as a currency string.
 *
 * @param amount       Numeric value (null/undefined → 'N/A')
 * @param currencyCode ISO 4217 code, e.g. 'AUD'
 * @param decimals     Decimal places (default: 0)
 * @param includeCode  Prefix with code label, e.g. 'AUD $123,456' (useful for
 *                     tables where multiple currencies might appear)
 */
export function formatCurrency(
  amount: number | null | undefined,
  currencyCode: string = DEFAULT_CURRENCY,
  decimals = 0,
  includeCode = false,
): string {
  if (amount == null) return 'N/A';
  const meta = getMeta(currencyCode);
  const formatted = new Intl.NumberFormat(meta.locale, {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  }).format(amount);
  const result = `${meta.symbol}${formatted}`;
  return includeCode ? `${currencyCode.toUpperCase()} ${result}` : result;
}

/**
 * Compact format for large numbers in summary cards: $1.2M, $345K, $12
 */
export function formatCurrencyCompact(
  amount: number | null | undefined,
  currencyCode: string = DEFAULT_CURRENCY,
): string {
  if (amount == null) return 'N/A';
  const meta = getMeta(currencyCode);
  const abs = Math.abs(amount);
  let value: number;
  let suffix: string;
  if (abs >= 1_000_000) {
    value = amount / 1_000_000;
    suffix = 'M';
  } else if (abs >= 1_000) {
    value = amount / 1_000;
    suffix = 'K';
  } else {
    return formatCurrency(amount, currencyCode, 0);
  }
  const formatted = new Intl.NumberFormat(meta.locale, {
    minimumFractionDigits: 0,
    maximumFractionDigits: 1,
  }).format(value);
  return `${meta.symbol}${formatted}${suffix}`;
}

/**
 * Format a plain number with locale-correct thousand separators.
 * Always uses en-AU format (123,456) regardless of browser locale.
 * Use this instead of .toLocaleString() to avoid Indian format (1,23,456).
 */
export function formatNumber(value: number | null | undefined): string {
  if (value == null) return '0';
  return new Intl.NumberFormat('en-AU').format(value);
}

/** Options list for the currency selector in the settings page. */
export const CURRENCY_OPTIONS = [
  { label: 'AUD — Australian Dollar ($)', value: 'AUD' },
  { label: 'USD — US Dollar ($)', value: 'USD' },
  { label: 'NZD — New Zealand Dollar ($)', value: 'NZD' },
  { label: 'GBP — British Pound (£)', value: 'GBP' },
  { label: 'EUR — Euro (€)', value: 'EUR' },
  { label: 'SGD — Singapore Dollar ($)', value: 'SGD' },
  { label: 'HKD — Hong Kong Dollar ($)', value: 'HKD' },
  { label: 'CAD — Canadian Dollar ($)', value: 'CAD' },
  { label: 'JPY — Japanese Yen (¥)', value: 'JPY' },
  { label: 'INR — Indian Rupee (₹)', value: 'INR' },
];